File: sdk/typescript/src/rpc.ts
import { createInterface } from "readline";

// JSON-RPC 2.0 types
export interface JsonRpcRequest {
  jsonrpc: "2.0";
  id: string | number;
  method: string;
  params?: Record<string, unknown>;
}

export interface JsonRpcResponse {
  jsonrpc: "2.0";
  id: string | number | null;
  result?: unknown;
  error?: JsonRpcError;
}

export interface JsonRpcNotification {
  jsonrpc: "2.0";
  method: string;
  params?: Record<string, unknown>;
}

export interface JsonRpcError {
  code: number;
  message: string;
  data?: unknown;
}

type JsonRpcMessage = JsonRpcRequest | JsonRpcResponse | JsonRpcNotification;

type RequestHandler = (
  params: Record<string, unknown>
) => Promise<unknown> | unknown;
type NotificationHandler = (params: Record<string, unknown>) => void;

/**
 * JSON-RPC 2.0 transport over stdin/stdout.
 */
export class RpcTransport {
  private nextId = 1;
  private pendingRequests = new Map<
    string | number,
    {
      resolve: (value: unknown) => void;
      reject: (error: Error) => void;
    }
  >();
  private requestHandlers = new Map<string, RequestHandler>();
  private notificationHandlers = new Map<string, NotificationHandler>();

  constructor() {
    const rl = createInterface({ input: process.stdin });
    rl.on("line", (line: string) => {
      if (!line.trim()) return;
      try {
        const message: JsonRpcMessage = JSON.parse(line);
        this.handleMessage(message);
      } catch {
        process.stderr.write(`Failed to parse JSON-RPC message: ${line}\n`);
      }
    });
    rl.on("close", () => process.exit(0));
  }

  /** Register a handler for incoming requests from the host. */
  onRequest(method: string, handler: RequestHandler): void {
    this.requestHandlers.set(method, handler);
  }

  /** Register a handler for incoming notifications from the host. */
  onNotification(method: string, handler: NotificationHandler): void {
    this.notificationHandlers.set(method, handler);
  }

  /** Send a request to the host and wait for a response. */
  async request(
    method: string,
    params?: Record<string, unknown>
  ): Promise<unknown> {
    const id = this.nextId++;
    const message: JsonRpcRequest = { jsonrpc: "2.0", id, method, params };
    this.send(message);

    return new Promise((resolve, reject) => {
      this.pendingRequests.set(id, { resolve, reject });
    });
  }

  /** Send a notification to the host (no response expected). */
  notify(method: string, params?: Record<string, unknown>): void {
    const message: JsonRpcNotification = { jsonrpc: "2.0", method, params };
    this.send(message);
  }

  private send(message: JsonRpcMessage): void {
    process.stdout.write(JSON.stringify(message) + "\n");
  }

  private handleMessage(message: JsonRpcMessage): void {
    if ("id" in message && "result" in message) {
      // Response
      this.handleResponse(message as JsonRpcResponse);
    } else if ("id" in message && "method" in message) {
      // Request
      this.handleRequest(message as JsonRpcRequest);
    } else if ("method" in message) {
      // Notification
      this.handleNotification(message as JsonRpcNotification);
    }
  }

  private handleResponse(response: JsonRpcResponse): void {
    const pending = this.pendingRequests.get(response.id!);
    if (!pending) return;
    this.pendingRequests.delete(response.id!);

    if (response.error) {
      pending.reject(
        new Error(`${response.error.message} (code: ${response.error.code})`)
      );
    } else {
      pending.resolve(response.result);
    }
  }

  private async handleRequest(request: JsonRpcRequest): Promise<void> {
    const handler = this.requestHandlers.get(request.method);
    if (!handler) {
      this.send({
        jsonrpc: "2.0",
        id: request.id,
        error: { code: -32601, message: `Method not found: ${request.method}` },
      } as JsonRpcResponse);
      return;
    }

    try {
      const result = await handler(request.params ?? {});
      this.send({
        jsonrpc: "2.0",
        id: request.id,
        result,
      } as JsonRpcResponse);
    } catch (err) {
      this.send({
        jsonrpc: "2.0",
        id: request.id,
        error: {
          code: -32603,
          message: err instanceof Error ? err.message : String(err),
        },
      } as JsonRpcResponse);
    }
  }

  private handleNotification(notification: JsonRpcNotification): void {
    const handler = this.notificationHandlers.get(notification.method);
    if (handler) {
      handler(notification.params ?? {});
    }
  }
}

File: sdk/python/sbx_plugin/rpc.py
"""JSON-RPC 2.0 transport over stdin/stdout."""

from __future__ import annotations

import asyncio
import json
import sys
from typing import Any, Callable


class RpcTransport:
    """Bidirectional JSON-RPC 2.0 over stdin/stdout."""

    def __init__(self):
        self._next_id = 1
        self._pending: dict[int | str, asyncio.Future] = {}
        self._notification_handlers: dict[str, Callable] = {}
        self._request_handlers: dict[str, Callable] = {}

    def on_notification(self, method: str, handler: Callable) -> None:
        self._notification_handlers[method] = handler

    def on_request(self, method: str, handler: Callable) -> None:
        self._request_handlers[method] = handler

    async def request(
        self, method: str, params: dict[str, Any] | None = None
    ) -> Any:
        msg_id = self._next_id
        self._next_id += 1
        message = {"jsonrpc": "2.0", "id": msg_id, "method": method}
        if params is not None:
            message["params"] = params
        self._send(message)

        loop = asyncio.get_event_loop()
        future = loop.create_future()
        self._pending[msg_id] = future
        result = await future
        return result

    def notify(self, method: str, params: dict[str, Any] | None = None) -> None:
        message: dict[str, Any] = {"jsonrpc": "2.0", "method": method}
        if params is not None:
            message["params"] = params
        self._send(message)

    def _send(self, message: dict) -> None:
        line = json.dumps(message, separators=(",", ":"))
        sys.stdout.write(line + "\n")
        sys.stdout.flush()

    async def _handle_message(self, message: dict) -> None:
        if "method" in message and "id" in message:
            # Request from host
            handler = self._request_handlers.get(message["method"])
            if handler:
                try:
                    result = handler(message.get("params", {}))
                    if asyncio.iscoroutine(result):
                        result = await result
                    response = {
                        "jsonrpc": "2.0",
                        "id": message["id"],
                        "result": result,
                    }
                except Exception as e:
                    response = {
                        "jsonrpc": "2.0",
                        "id": message["id"],
                        "error": {"code": -32603, "message": str(e)},
                    }
                self._send(response)
        elif "method" in message:
            # Notification from host
            handler = self._notification_handlers.get(message["method"])
            if handler:
                try:
                    result = handler(message.get("params", {}))
                    if asyncio.iscoroutine(result):
                        await result
                except Exception as e:
                    print(f"Notification handler error: {e}", file=sys.stderr)
        elif "id" in message:
            # Response to our request
            future = self._pending.pop(message["id"], None)
            if future and not future.done():
                if "error" in message:
                    err = message["error"]
                    future.set_exception(
                        RuntimeError(f"{err['message']} (code: {err['code']})")
                    )
                else:
                    future.set_result(message.get("result"))

    async def _read_loop(self) -> None:
        loop = asyncio.get_event_loop()
        reader = asyncio.StreamReader()
        protocol = asyncio.StreamReaderProtocol(reader)
        await loop.connect_read_pipe(lambda: protocol, sys.stdin)

        while True:
            line = await reader.readline()
            if not line:
                break
            try:
                message = json.loads(line.decode())
                await self._handle_message(message)
            except json.JSONDecodeError:
                print(f"Failed to parse: {line}", file=sys.stderr)

    def run(self) -> None:
        """Start the event loop and read messages until stdin closes."""
        asyncio.run(self._read_loop())
